src/Ogooreck.Sample.BusinessLogic.FSharp.Tests/Deciders/BankAccountDecider.fs

```diff
@@ -0,0 +1,74 @@
+﻿module Deciders.BankAccountDecider
+
+open System
+open BankAccount
+open Deciders.BankAccount
+
+type OpenBankAccount =
+    { BankAccountId: Guid
+      AccountNumber: string
+      ClientId: Guid
+      CurrencyISOCode: string }
+
+type RecordDeposit = { Amount: decimal; CashierId: Guid }
+type WithdrawnCashFromATM = { Amount: decimal; AtmId: Guid }
+type CloseBankAccount = { Reason: string }
+
+type Command =
+    | OpenBankAccount of OpenBankAccount
+    | RecordDeposit of RecordDeposit
+    | WithdrawnCashFromATM of WithdrawnCashFromATM
+    | CloseBankAccount of CloseBankAccount
+
+let openBankAccount now command =
+    { BankAccountId = command.BankAccountId
+      AccountNumber = command.AccountNumber
+      ClientId = command.ClientId
+      CurrencyISOCode = command.CurrencyISOCode
+      CreatedAt = now ()
+      Version = 1 }
+
+let recordDeposit now (command: RecordDeposit, bankAccount) =
+    if (bankAccount.Status = BankAccountStatus.Closed) then
+        invalidOp "Account is closed!"
+
+    { BankAccountId = bankAccount.Id
+      Amount = command.Amount
+      CashierId = command.CashierId
+      RecordedAt = now ()
+      Version = bankAccount.Version + 1L }
+
+let withdrawCashFromATM now (command, bankAccount) =
+    if (bankAccount.Status = BankAccountStatus.Closed) then
+        invalidOp "Account is closed!"
+
+    if (bankAccount.Balance < command.Amount) then
+        invalidOp "Account is closed!"
+
+    { BankAccountId = bankAccount.Id
+      Amount = command.Amount
+      ATMId = command.AtmId
+      RecordedAt = now ()
+      Version = bankAccount.Version + 1L }
+
+let closeBankAccount now (command, bankAccount) =
+    if (bankAccount.Status = BankAccountStatus.Closed) then
+        invalidOp "Account is already closed!"
```

Comment:
What about encoding invalid op in the form of an event, e.g. `BankAccountAlreadyClosed`?

Reply:
I'd do it if I'd like to have a history of wrong operations. Or in Saga, where I'd like to return the failure event. In this example, I assume that's something that should not happen, thus throwing exceptions (per what I found in https://fsharpforfunandprofit.com/posts/against-railway-oriented-programming/).

Reply:
idempotency FTW ;)

Reply:
Yes, the point is:
* If the rest of the system ensure it should not happen, then you can throw an exception. When it happens there is a bug in the system
* If it will happen whatever you do and you have to do things when it happens, then use an event.